src/opaque.rs

```diff
@@ -0,0 +1,178 @@
+use std::{
+    io::{Cursor, Write},
+    marker::PhantomData,
+};
+
+use byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};
+
+use crate::Error;
+
+// Opaque is a Variable-length Array that holds an uninterpreted byte array
+//https://datatracker.ietf.org/doc/html/rfc1014#section-3.12
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct Opaque<'a, T>
+where
+    T: AsRef<[u8]>,
+{
+    body: T,
+    phantom: PhantomData<&'a T>,
+}
```

Comment:
This might be causing you a bit of a headache!

I think you're probably after something like this:

```suggestion
pub struct Opaque<T> {
    body: T,
}
```

Which says "any buffer, with any lifetime". This is different from what is there currently, that says "any buffer (possibly borrowed), but pretend `Opaque` also contains some additional reference" (that's the `PhantomData<&'a T>` at play).

Additionally we don't bound the type, but rather the impl only, which prevents it from "leaking" into every type that encapsulates this type.

After that, the `TryFrom` impl will be something like this:

```rust
impl<'a> TryFrom<&mut Cursor<&'a [u8]>> for Opaque<&'a [u8]> {
    type Error = Error;

    /// Deserialises a new [`Opaque`] from `cursor`.
    fn try_from(c: &mut Cursor<&'a [u8]>) -> Result<Self, Self::Error> {
        let len = c.read_u32::<BigEndian>()?;
        let data = *c.get_ref();
        let start = c.position() as usize;
        let end = start + len as usize;
        let padded_end = pad_length(len) + end as u32;

        c.set_position(padded_end as u64);
        Ok(Opaque {
            body: &data[start..end],
            // phantom: PhantomData,
        })
    }
}
```

Here we're saying "the T is a borrowed byte slice" and importantly the T contains the lifetime, not the `Opaque` itself (a bit confusing, I know!)

This should mean there's a lot less lifetimes to faff with, and it'll be a bit easier to integrate with `bytes` too!